### ISSUE TITLE
Problem on block device rate-limiting

### ISSUE THREAD

**Issue by namespaceXP** [firecracker-microvm/firecracker#2176]
I'm interested in I/O throttling. While testing the rate-limiter on block device, I use the following setting to set the average iops to 3000: 

"rate_limiter": {
  "ops": {
    "size": 3000,
    "one_time_burst": 0,
    "refill_time": 1000
    }
}

And use fio for testing the iops: 

`fio -rw=randwrite -ioengine=libaio -bs=4k -size=10G -numjobs=1 -runtime=1000 -group_reporting -filename=iotest -name=Rand_Write_Testing -direct=1 -iodepth=128`

However, I've found that the average iops is about 2765, somehow lower than expected(3000). I wonder the reason.

**Comment by raduiliescu** [comment]
Before diving deep on the configuration, what performance number do you get without rate-limiter?

**Comment by namespaceXP** [comment]
> Before diving deep on the configuration, what performance number do you get without rate-limiter?

The average IOPS without rate limiter is about 56000.

**Comment by serban300** [comment]
This might be happening because the rate limiter also counts Flush and DeviceID requests. I will check it.

**Comment by serban300** [comment]
I did some more research here. This is not caused by Flush and DeviceID requests. This seems to be the desired behavior for the rate limiter. 

The rate limiter allows 3000 OPS, after which it waits 1000 ms and then it receives a new budget of 3000 OPS. So the configuration that you used for the rate limiter doesn't translate to 3000 IOPS. it translates to cycles like the following:
```
+--------------------------+-------------+
| 3000 OPS                 | refill time |
+--------------------------+-------------+
| time needed for 3000 OPS | 1000 ms     |
+--------------------------+-------------+
```

In your case the time needed for 3000 OPS is approximately 1000 ms / 56000 * 3000 = 54ms . So for example in 1 minute there will be 60000 / 1054 = 57 such cycles. 57 * 3000 / 60 = 2850 IOPS which is close to the result that you are getting.

@namespaceXP does this answear your question ? Do you have any concerns related to this behavior ?

@acatangiu It seems that you implemented the block device rate limiting. Can you confirm if this is the desired behavior ?

**Comment by acatangiu** [comment]
@namespaceXP The firecracker rate-limiter implementation leans more on efficiency at the expense of accuracy.

It is a _passive_ implementation driven by the consumer (in this case, disk IOPS). Configured ratelimiter uses a token bucket of `3000` with a `1000ms` refill rate, which should lead to exactly `3000 IOPS/s`, but because we don't want to incur the cost of floating point operations the resulting rate is always lower.

Short summary of how it works:
Each time `x` tokens go through limiter (for block iops, it will always be **1 io-op** trying to go through) the rate-limiter:
1. calculates how much time has passed since last operation
2. replenishes its budget based on elapsed time above
3. attempts to allow `X` if it fits the budget
    1. if it does fit, `X` is subtracted from budget and operation succeeds (block IO op is done)
    2. if not, operation is postponed for `100ms` (_this is a hardcoded timer delay unrelated to `refill rate` - simplicity and efficiency again_), then is retried from the top.

Now the _problem_ you are seeing is most likely caused by `step 2` described above, where non-floating point division loses precision as can be seen [here](https://github.com/firecracker-microvm/firecracker/blob/ffbc2efcf2dcffa02db6df3474a8ad2b0bad21f7/src/rate_limiter/src/lib.rs#L175):
```
// At each 'time_delta' nanoseconds the bucket should refill with:
// refill_amount = (time_delta * size) / (complete_refill_time_ms * 1_000_000)
```

So between two _block device IO operations_ `Y` nanoseconds pass meaning `refill_amount` would have to replenish with:
`Y * 3_000 / 1_000 * 1_000_000`; so it becomes apparent that there is precision loss and rounding-down on `refill_amount` leading to overall lower rate than configured.

This is even more apparent for sub-millisecond values for `Y`, where the result there is `0`. This is why this disparity is most obvious in synthetic benchmarks where a `firecracker` microvm does little else than focus on those ops, resulting in small and often `Y`s.

----------

Now, we are aware of this behavior and know of a few ways to fix it. We haven't done so yet because afaik all fixes we thought of are either costly/inefficient (floating point ops for example) or add considerable complexity to the rate-limiter implementation (like some clever maths to minimize precision loss).

So far, we have not had any complaints from real-world customer production workloads so it was never prioritized.

Is your usecase negatively impacted by this? If so, can you describe your requirements and setup, maybe we can offer tuning options to unblock you?
